packages/gatsby/src/schema/infer/__tests__/infer.js

```diff
@@ -1053,6 +1053,39 @@ Object {
         slash(path.resolve(dir, `file_2.txt`))
       )
     })
+
+    it(`Links to file node from non-standard field name`, async () => {
+      const fieldWithSpecialChars = `file-ж-ä-!@#$%^&*()_-=+:;'"?,~\``
+      const nodes = [
+        {
+          id: `1`,
+          "file-dashed": `./file_1.jpg`,
+          [fieldWithSpecialChars]: `./file_1.jpg`,
+          parent: `parent`,
+          internal: { type: `Test` },
+        },
+      ].concat(getFileNodes())
+
+      const result = await getQueryResult(
+        nodes,
+        `
+          file_dashed {
+            absolutePath
+          }
+          file___________________________ {
```

Comment:
I think this PR should be reviewed by someone else; but out of curiosity, what's with the underscores?

Reply:
We replace special chars found in node fields with underscores in GraphQL schema. So this is basically this line converted to GraphQL-compatible name:

https://github.com/gatsbyjs/gatsby/blob/67aa4baf9c8e4713c38a08cd0f4aa2241949ccc4/packages/gatsby/src/schema/infer/__tests__/infer.js#L1058

This is expected behavior (works the same before this PR). The issue is that before this PR the reverse link from GraphQL field to Gatsby node field was lost.

Reply:
Ahhh, makes sense. Thanks for the explanation!

Reply:
This is hilarious 😆

Reply:
but is this replacement not a problem with 2 similar fields like:

- `file-Nüsse` -> `file_N_sse`
- `file-Nässe` -> `file_N_sse`

which both results in same `file_N_sse`?

maybe it makes sense to encode the names with [Punycode](https://en.wikipedia.org/wiki/Punycode) which results in:

- `file-Nüsse` -> `xn--file-nsse-v9a`
- `file-Nässe` -> `xn--file-nsse-02a`

and also works for normal code:

- `file-Nasse` -> `file-Nasse`

Reply:
@muescha That is a good point! 

@vladar Thoughts?

Reply:
one drawback: 

it is not anymore nice readable 
- but maybe it is possible to add an field `original_field_name` where the name can shown in the query result
- or an comment at the schema doc with the original name can be somehow generated?

Reply:
ok - one more drawback: 
- punycode results in `[a-z0-9-]` but then a replacement from `-` to `_` is possible...

Reply:
I agree that we could do better here but this is out of the scope of this PR. Also changing this behavior would be breaking.

Anyways, I tried finding any open issue for something like this but it looks like there are none, which means it is not that big of a problem. Of course, we can open a separate issue to handle this but I feel that it would be much better to have an issue from a person who really suffers from this problem vs. us doing this.

As for the technical part - we could use slugify as we already do to generate query name from the file path:

https://github.com/gatsbyjs/gatsby/blob/21d6a874d506f253767cfbef4deb0eb3618ef743/packages/gatsby/src/query/file-parser.js#L28-L40